core/dbt/events/base_types.py

```diff
@@ -0,0 +1,102 @@
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from datetime import datetime
+import os
+from typing import Any
+
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# These base types define the _required structure_ for the concrete event #
+# types defined in types.py                                               #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+
+# in preparation for #3977
+class TestLevel():
+    def level_tag(self) -> str:
+        return "test"
+
+
+class DebugLevel():
+    def level_tag(self) -> str:
+        return "debug"
+
+
+class InfoLevel():
+    def level_tag(self) -> str:
+        return "info"
+
+
+class WarnLevel():
+    def level_tag(self) -> str:
+        return "warn"
+
+
+class ErrorLevel():
+    def level_tag(self) -> str:
+        return "error"
+
+
+@dataclass
+class ShowException():
+    # N.B.:
+    # As long as we stick with the current convention of setting the member vars in the
+    # `message` method of subclasses, this is a safe operation.
+    # If that ever changes we'll want to reassess.
+    def __post_init__(self):
+        self.exc_info: Any = True
+        self.stack_info: Any = None
+        self.extra: Any = None
+
+
+# TODO add exhaustiveness checking for subclasses
+# top-level superclass for all events
+class Event(metaclass=ABCMeta):
+    # fields that should be on all events with their default implementations
+    ts: datetime = datetime.now()
+    pid: int = os.getpid()
+
+    # four digit string code that uniquely identifies this type of event
+    # uniqueness and valid characters are enforced by tests
+    @classmethod
+    @abstractmethod
+    def code(cls) -> str:
+        raise Exception("code() not implemented for event")
```

Comment:
Here's the abstract method that's doing the enforcement via mypy. I couldn't figure out how to make mypy check that subclasses have defined an abstract property even though the internet seems to think you can do that. You'll need to do one of two things to get this to work:
1. figure out how to make an abstract property such that if you delete `code` from one of the concrete classes, mypy fails.
2. change all the codes from fields to functions that override this one in each of the concrete classes.

Reply:
I just tried this by commenting out the code on one of the concrete tests and running mypy, and it worked. I'm not sure why though. @iknox-fa do you know why a @dataclass field named code in the concrete classes counts as overriding this abstract method??

Reply:
Because it's now an abstract property.... and data fields are just automagic properties generated by an automagic `__init__` method.

Reply:
too implicit for my liking, but I'm glad it works.